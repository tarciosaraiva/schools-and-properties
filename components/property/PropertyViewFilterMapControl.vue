<template>
  <div>
    <div class="d-flex align-items-center justify-content-between">
      <span>{{ propertyTypes }}</span>
      <a href="#" class="pl-button small link" @click.prevent.stop="openPropertyFilter">
        <font-awesome-icon icon="sliders-h"/>
      </a>
    </div>
    <div class="d-flex align-items-center justify-content-between">
      <font-awesome-icon icon="bed"/>
      {{ propertiesFilter.rooms }}
      <font-awesome-icon icon="bath"/>
      {{ propertiesFilter.bathrooms }}
      <font-awesome-icon icon="car"/>
      {{ propertiesFilter.carSpaces }}
      <font-awesome-icon icon="dollar-sign"/>
      {{ fmtPrice }}
      <a href="#" class="pl-button small link" @click.prevent.stop="loadListings">
        <font-awesome-icon icon="sync"/>
      </a>
    </div>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import { mapGetters, mapActions } from 'vuex'

export default Vue.extend({
  props: {
    openPropertyFilter: Function
  },
  computed: {
    ...mapGetters(['propertiesFilter']),
    fmtPrice () {
      return new Intl.NumberFormat('en-AU').format(this.propertiesFilter.maxPrice)
    },
    propertyTypes () {
      return this.propertiesFilter.types.join(', ')
    }
  },
  methods: {
    ...mapActions(['loadListings']),
  }
})
</script>
