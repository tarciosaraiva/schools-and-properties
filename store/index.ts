import { GetterTree, ActionTree, MutationTree } from 'vuex'

interface ApplicationFilter {
  properties: PropertiesFilter,
  schools: SchoolsFilter
}

interface PropertiesFilter {
  types: string[],
  rooms: number,
  bathrooms: number,
  carSpaces: number,
  maxPrice: number
}

export interface SchoolsFilter {
  educationSector: string,
  zone: string,
  primary: SchoolLevelRating,
  secondary: SchoolLevelRating,
}

interface SchoolLevelRating {
  plot: boolean,
  rating: number,
  englishRating: number,
  mathsRating: number
}

export interface PropertyListing {
  id: number,
  price: string,
  area: string
  bathrooms: number,
  bedrooms: number,
  carspaces: number,
  displayableAddress: string,
  features: string[],
  landArea: number,
  latitude: number,
  longitude: number,
  postcode: string,
  propertyType: string
  region: string,
  state: string,
  street: string,
  streetNumber: string,
  suburb: string,
  unitNumber: string,
  media: string[],
  listingSlug: string
}

export interface School {
  schoolName: string,
  educationSector: string,
  schoolType: string,
  phoneNumber: string,
  primaryOverallScore: string,
  primaryEnglishScore: string,
  primaryMathsScore: string,
  primaryEnrolments: string,
  secondaryOverallScore: string,
  secondaryEnglishScore: string,
  secondaryMathsScore: string,
  secondaryEnrolments: string
}

export const state = () => ({
  bounds: {
    ne: { lat: 0, lon: 0 },
    nw: { lat: 0, lon: 0 },
    se: { lat: 0, lon: 0 },
    sw: { lat: 0, lon: 0 },
  },
  geocodeSuggestions: [] as any,
  list: [] as PropertyListing[],
  filter: {
    properties: {
      types: ['House', 'Townhouse'],
      rooms: 3,
      bathrooms: 2,
      carSpaces: 1,
      maxPrice: 900000
    },
    schools: {
      educationSector: 'all',
      zone: 'none',
      primary: {
        plot: true,
        rating: 95,
        englishRating: 4,
        mathsRating: 4
      },
      secondary: {
        plot: true,
        rating: 95,
        englishRating: 4,
        mathsRating: 4
      }
    }
  } as ApplicationFilter
})

export type RootState = ReturnType<typeof state>

export const getters: GetterTree<RootState, RootState> = {
  listings: state => state.list,
  schoolsFilter: state => state.filter.schools,
  propertiesFilter: state => state.filter.properties,
  locations: state => state.geocodeSuggestions,
  listingPois: state => state.list.map(l => ({
    type: 'Feature',
    id: l.id,
    geometry: {
      type: 'Point',
      coordinates: [l.longitude, l.latitude]
    },
    properties: { ...l }
  }))
}

export const mutations: MutationTree<RootState> = {
  ADD_LISTINGS (state, listings: any[]) {
    const propertyListings = listings
      .filter(l => l.type === 'PropertyListing')
      .map(({ listing }) => {
        const propertyListing: PropertyListing = {
          id: listing.id,
          price: listing.priceDetails?.displayPrice,
          area: listing.propertyDetails.area,
          bathrooms: listing.propertyDetails.bathrooms,
          bedrooms: listing.propertyDetails.bedrooms,
          carspaces: listing.propertyDetails.carspaces,
          displayableAddress: listing.propertyDetails.displayableAddress,
          features: listing.propertyDetails.features,
          landArea: listing.propertyDetails.landArea,
          latitude: listing.propertyDetails.latitude,
          longitude: listing.propertyDetails.longitude,
          postcode: listing.propertyDetails.postcode,
          propertyType: listing.propertyDetails.propertyType,
          region: listing.propertyDetails.region,
          state: listing.propertyDetails.state,
          street: listing.propertyDetails.street,
          streetNumber: listing.propertyDetails.streetNumber,
          suburb: listing.propertyDetails.suburb,
          unitNumber: listing.propertyDetails.unitNumber,
          media: listing.media.map((m: any) => m.url),
          listingSlug: listing.listingSlug
        }

        return propertyListing
      })

    state.list = propertyListings
  },
  UPDATE_PROPERTY_FILTER_TYPE (state, payload) {
    state.filter.properties.types = payload
  },
  UPDATE_PROPERTY_FILTER_ROOMS (state, payload) {
    state.filter.properties.rooms = Number(payload)
  },
  UPDATE_PROPERTY_FILTER_BATHROOMS (state, payload) {
    state.filter.properties.bathrooms = Number(payload)
  },
  UPDATE_PROPERTY_FILTER_CAR_SPACES (state, payload) {
    state.filter.properties.carSpaces = Number(payload)
  },
  UPDATE_PROPERTY_FILTER_MAX_PRICE (state, payload) {
    state.filter.properties.maxPrice = Number(payload)
  },
  UPDATE_BOUNDING_BOX (state, payload) {
    state.bounds.nw.lat = payload.getNorthWest().lat
    state.bounds.nw.lon = payload.getNorthWest().lng
    state.bounds.ne.lat = payload.getNorthEast().lat
    state.bounds.ne.lon = payload.getNorthEast().lng
    state.bounds.se.lat = payload.getSouthEast().lat
    state.bounds.se.lon = payload.getSouthEast().lng
    state.bounds.sw.lat = payload.getSouthWest().lat
    state.bounds.sw.lon = payload.getSouthWest().lng
  },
  UPDATE_SCHOOL_EDUCATION_SECTOR (state, payload) {
    state.filter.schools.educationSector = payload
  },
  UPDATE_SCHOOL_ZONE (state, payload) {
    state.filter.schools.zone = payload
  },
  UPDATE_PLOT_PRIMARY_SCHOOLS (state, payload) {
    state.filter.schools.primary.plot = payload
  },
  UPDATE_SCHOOL_PRIMARY_RATING (state, payload) {
    state.filter.schools.primary.rating = Number(payload)
  },
  UPDATE_SCHOOL_PRIMARY_ENG_RATING (state, payload) {
    state.filter.schools.primary.englishRating = Number(payload)
  },
  UPDATE_SCHOOL_PRIMARY_MATH_RATING (state, payload) {
    state.filter.schools.primary.mathsRating = Number(payload)
  },
  UPDATE_PLOT_SECONDARY_SCHOOLS (state, payload) {
    state.filter.schools.secondary.plot = payload
  },
  UPDATE_SCHOOL_SECONDARY_RATING (state, payload) {
    state.filter.schools.secondary.rating = Number(payload)
  },
  UPDATE_SCHOOL_SECONDARY_ENG_RATING (state, payload) {
    state.filter.schools.secondary.englishRating = Number(payload)
  },
  UPDATE_SCHOOL_SECONDARY_MATH_RATING (state, payload) {
    state.filter.schools.secondary.mathsRating = Number(payload)
  },
  ADD_GEOCODE_RESULTS (state, payload) {
    state.geocodeSuggestions = payload.map((f: any) => ({
      center: f.center,
      name: f.place_name,
      placeType: f.place_type
    }))
  }
}

export const actions: ActionTree<RootState, RootState> = {
  async loadListings ({ commit, state }) {
    const { filter, bounds } = state

    const payload = {
      listingType: 'Sale',
      pageSize: 200,
      propertyTypes: filter.properties.types,
      minBedrooms: filter.properties.rooms,
      minBathrooms: filter.properties.bathrooms,
      minCarspaces: filter.properties.carSpaces,
      maxPrice: filter.properties.maxPrice,
      locations: [{ state: 'VIC' }],
      geoWindow: { box: { topLeft: bounds.nw, bottomRight: bounds.se } }
    }

    const listings = await this.$axios.$post('https://api.domain.com.au/v1/listings/residential/_search', payload, { headers: {'X-Api-Key': process.env.domainApiKey } })
    commit('ADD_LISTINGS', listings)
  },

  async geocode ({ commit }, searchTerm) {
    if (searchTerm) {
      const params = {
        key: process.env.mapTilerSecret,
        bbox: '141.24847451171752,-39.624732230379166,148.68072548828115,-34.32302549095252',
        language: 'en'
      }

      const geocodeResults = await this.$axios.$get(`https://api.maptiler.com/geocoding/${searchTerm}.json`, { params })
      const schoolsResults = await this.$axios.$get(`//${process.env.domain}/schools/schools.json`)
      const mappedSchoolResults = schoolsResults.map((s: any) => ({
        center: [s.lon, s.lat],
        place_name: s.schoolName,
        place_type: 'school'
      })).filter((s: any) => s.place_name.toLowerCase().includes(searchTerm))

      commit('ADD_GEOCODE_RESULTS', [...geocodeResults.features, ...mappedSchoolResults])
    } else {
      commit('ADD_GEOCODE_RESULTS', [])
    }
  },

  setBoundingBox({ commit }, payload) {
    commit('UPDATE_BOUNDING_BOX', payload)
  }
}
