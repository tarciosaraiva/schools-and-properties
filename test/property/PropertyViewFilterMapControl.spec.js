import { mount, createLocalVue } from '@vue/test-utils'
import { library } from '@fortawesome/fontawesome-svg-core'
import { fas } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
import Vuex from 'vuex'

import { state, getters } from '@/store'
import PropertyViewFilterMapControl from '@/components/property/PropertyViewFilterMapControl.vue'

library.add(fas)

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.component('font-awesome-icon', FontAwesomeIcon)

describe('PropertyViewFilterMapControl', () => {
  const propsData = {
    openPropertyFilter: jest.fn()
  }

  let store
  let actions

  beforeEach(() => {
    actions = {
      loadListings: jest.fn()
    }
    store = new Vuex.Store({ state, getters, actions })
  })

  test('is a Vue instance', () => {
    const wrapper = mount(PropertyViewFilterMapControl, { propsData, store, localVue })
    expect(wrapper.vm).toBeTruthy()
  })

  test('renders correctly with props', () => {
    const wrapper = mount(PropertyViewFilterMapControl, { propsData, store, localVue })
    expect(wrapper).toMatchSnapshot()
  })

  test('should open property filter modal when clicking on settings button', () => {
    const wrapper = mount(PropertyViewFilterMapControl, { propsData, store, localVue })
    const links = wrapper.findAll('a')
    links.at(0).trigger('click')
    expect(propsData.openPropertyFilter).toHaveBeenCalled()
  })

  test('should refresh listing data when clicking on sync button', () => {
    const wrapper = mount(PropertyViewFilterMapControl, { propsData, store, localVue })
    const links = wrapper.findAll('a')
    links.at(1).trigger('click')
    expect(actions.loadListings).toHaveBeenCalled()
  })
})
