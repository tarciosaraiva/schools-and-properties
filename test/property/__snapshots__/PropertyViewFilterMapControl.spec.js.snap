// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyViewFilterMapControl renders correctly with props 1`] = `
<div>
  <div class="d-flex align-items-center justify-content-between">
    <span>House, Townhouse</span>
    <a class="pl-button small link" href="#">
      <svg aria-hidden="true" class="svg-inline--fa fa-sliders-h fa-w-16" data-icon="sliders-h" data-prefix="fas" focusable="false" role="img" viewBox="0 0 512 512" xmlns="http://www.w3.org/2000/svg">
        <path class="" d="M496 384H160v-16c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v16H16c-8.8 0-16 7.2-16 16v32c0 8.8 7.2 16 16 16h80v16c0 8.8 7.2 16 16 16h32c8.8 0 16-7.2 16-16v-16h336c8.8 0 16-7.2 16-16v-32c0-8.8-7.2-16-16-16zm0-160h-80v-16c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v16H16c-8.8 0-16 7.2-16 16v32c0 8.8 7.2 16 16 16h336v16c0 8.8 7.2 16 16 16h32c8.8 0 16-7.2 16-16v-16h80c8.8 0 16-7.2 16-16v-32c0-8.8-7.2-16-16-16zm0-160H288V48c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v16H16C7.2 64 0 71.2 0 80v32c0 8.8 7.2 16 16 16h208v16c0 8.8 7.2 16 16 16h32c8.8 0 16-7.2 16-16v-16h208c8.8 0 16-7.2 16-16V80c0-8.8-7.2-16-16-16z" fill="currentColor" />
      </svg>
    </a>
  </div>
  <div class="d-flex align-items-center justify-content-between">
    <svg aria-hidden="true" class="svg-inline--fa fa-bed fa-w-20" data-icon="bed" data-prefix="fas" focusable="false" role="img" viewBox="0 0 640 512" xmlns="http://www.w3.org/2000/svg">
      <path class="" d="M176 256c44.11 0 80-35.89 80-80s-35.89-80-80-80-80 35.89-80 80 35.89 80 80 80zm352-128H304c-8.84 0-16 7.16-16 16v144H64V80c0-8.84-7.16-16-16-16H16C7.16 64 0 71.16 0 80v352c0 8.84 7.16 16 16 16h32c8.84 0 16-7.16 16-16v-48h512v48c0 8.84 7.16 16 16 16h32c8.84 0 16-7.16 16-16V240c0-61.86-50.14-112-112-112z" fill="currentColor" />
    </svg>
    3
    <svg aria-hidden="true" class="svg-inline--fa fa-bath fa-w-16" data-icon="bath" data-prefix="fas" focusable="false" role="img" viewBox="0 0 512 512" xmlns="http://www.w3.org/2000/svg">
      <path class="" d="M32,384a95.4,95.4,0,0,0,32,71.09V496a16,16,0,0,0,16,16h32a16,16,0,0,0,16-16V480H384v16a16,16,0,0,0,16,16h32a16,16,0,0,0,16-16V455.09A95.4,95.4,0,0,0,480,384V336H32ZM496,256H80V69.25a21.26,21.26,0,0,1,36.28-15l19.27,19.26c-13.13,29.88-7.61,59.11,8.62,79.73l-.17.17A16,16,0,0,0,144,176l11.31,11.31a16,16,0,0,0,22.63,0L283.31,81.94a16,16,0,0,0,0-22.63L272,48a16,16,0,0,0-22.62,0l-.17.17c-20.62-16.23-49.83-21.75-79.73-8.62L150.22,20.28A69.25,69.25,0,0,0,32,69.25V256H16A16,16,0,0,0,0,272v16a16,16,0,0,0,16,16H496a16,16,0,0,0,16-16V272A16,16,0,0,0,496,256Z" fill="currentColor" />
    </svg>
    2
    <svg aria-hidden="true" class="svg-inline--fa fa-car fa-w-16" data-icon="car" data-prefix="fas" focusable="false" role="img" viewBox="0 0 512 512" xmlns="http://www.w3.org/2000/svg">
      <path class="" d="M499.99 176h-59.87l-16.64-41.6C406.38 91.63 365.57 64 319.5 64h-127c-46.06 0-86.88 27.63-103.99 70.4L71.87 176H12.01C4.2 176-1.53 183.34.37 190.91l6 24C7.7 220.25 12.5 224 18.01 224h20.07C24.65 235.73 16 252.78 16 272v48c0 16.12 6.16 30.67 16 41.93V416c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32v-32h256v32c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32v-54.07c9.84-11.25 16-25.8 16-41.93v-48c0-19.22-8.65-36.27-22.07-48H494c5.51 0 10.31-3.75 11.64-9.09l6-24c1.89-7.57-3.84-14.91-11.65-14.91zm-352.06-17.83c7.29-18.22 24.94-30.17 44.57-30.17h127c19.63 0 37.28 11.95 44.57 30.17L384 208H128l19.93-49.83zM96 319.8c-19.2 0-32-12.76-32-31.9S76.8 256 96 256s48 28.71 48 47.85-28.8 15.95-48 15.95zm320 0c-19.2 0-48 3.19-48-15.95S396.8 256 416 256s32 12.76 32 31.9-12.8 31.9-32 31.9z" fill="currentColor" />
    </svg>
    1
    <svg aria-hidden="true" class="svg-inline--fa fa-dollar-sign fa-w-9" data-icon="dollar-sign" data-prefix="fas" focusable="false" role="img" viewBox="0 0 288 512" xmlns="http://www.w3.org/2000/svg">
      <path class="" d="M209.2 233.4l-108-31.6C88.7 198.2 80 186.5 80 173.5c0-16.3 13.2-29.5 29.5-29.5h66.3c12.2 0 24.2 3.7 34.2 10.5 6.1 4.1 14.3 3.1 19.5-2l34.8-34c7.1-6.9 6.1-18.4-1.8-24.5C238 74.8 207.4 64.1 176 64V16c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v48h-2.5C45.8 64-5.4 118.7.5 183.6c4.2 46.1 39.4 83.6 83.8 96.6l102.5 30c12.5 3.7 21.2 15.3 21.2 28.3 0 16.3-13.2 29.5-29.5 29.5h-66.3C100 368 88 364.3 78 357.5c-6.1-4.1-14.3-3.1-19.5 2l-34.8 34c-7.1 6.9-6.1 18.4 1.8 24.5 24.5 19.2 55.1 29.9 86.5 30v48c0 8.8 7.2 16 16 16h32c8.8 0 16-7.2 16-16v-48.2c46.6-.9 90.3-28.6 105.7-72.7 21.5-61.6-14.6-124.8-72.5-141.7z" fill="currentColor" />
    </svg>
    900,000
    <a class="pl-button small link" href="#">
      <svg aria-hidden="true" class="svg-inline--fa fa-sync fa-w-16" data-icon="sync" data-prefix="fas" focusable="false" role="img" viewBox="0 0 512 512" xmlns="http://www.w3.org/2000/svg">
        <path class="" d="M440.65 12.57l4 82.77A247.16 247.16 0 0 0 255.83 8C134.73 8 33.91 94.92 12.29 209.82A12 12 0 0 0 24.09 224h49.05a12 12 0 0 0 11.67-9.26 175.91 175.91 0 0 1 317-56.94l-101.46-4.86a12 12 0 0 0-12.57 12v47.41a12 12 0 0 0 12 12H500a12 12 0 0 0 12-12V12a12 12 0 0 0-12-12h-47.37a12 12 0 0 0-11.98 12.57zM255.83 432a175.61 175.61 0 0 1-146-77.8l101.8 4.87a12 12 0 0 0 12.57-12v-47.4a12 12 0 0 0-12-12H12a12 12 0 0 0-12 12V500a12 12 0 0 0 12 12h47.35a12 12 0 0 0 12-12.6l-4.15-82.57A247.17 247.17 0 0 0 255.83 504c121.11 0 221.93-86.92 243.55-201.82a12 12 0 0 0-11.8-14.18h-49.05a12 12 0 0 0-11.67 9.26A175.86 175.86 0 0 1 255.83 432z" fill="currentColor" />
      </svg>
    </a>
  </div>
</div>
`;
